subsys/bluetooth/host/gatt.c

```diff
@@ -2011,44 +2013,46 @@ static u8_t update_ccc(const struct bt_gatt_attr *attr, void *user_data)
 
 	ccc = attr->user_data;
 
-	for (i = 0; i < ARRAY_SIZE(ccc->cfg); i++) {
-		/* Ignore configuration for different peer */
-		if (bt_conn_addr_le_cmp(conn, &ccc->cfg[i].peer)) {
-			continue;
+	/* Check if attribute requires encryption/authentication */
+	need_update = false;
+	err = bt_gatt_check_perm(conn, attr, BT_GATT_PERM_WRITE_MASK);
```

Comment:
Im don't really like this sort of change since this is altering the order to call bt_gatt_check_perm before this only pays off when there is more than one peer subscribed and will introduce a call to bt_gatt_check_perm even for attributes that have never been subscribed to, so Id say it is a bad trade since it makes the code less readable.

Reply:
@wopu-ot Do you have a measurement of performance gain by this PR? Is it significant?

Reply:
Calls to `bt_gatt_check_perm` are cheaper than calls to `bt_conn_addr_le_cmp`  and and the loop iterates over the whole array of peers even if there are no subscriptions. This will almost always be beneficial. Obviously, avoiding the linear search for the peer in the first place would be better, but that would be a much larger change.
In terms of performance gain, I measured `bt_gatt_connected` and `bt_gatt_encrypt_change` at ~2.5ms with the original code and at ~2ms with the updated code.

Reply:
We could basically just make the check for cfg->value before the call to bt_conn_addr_le_cmp so we only do call bt_conn_addr_le_cmp and bt_gatt_check_perm if it is configured, then it will most likely be the same or even lower is we don't have to call into both function if value is not set (which is probably can be considered a bug since that would bump security even when no subscription would have been found).

Reply:
```
diff --git a/subsys/bluetooth/host/gatt.c b/subsys/bluetooth/host/gatt.c
index 378f4eb004..1ce6868b77 100644
--- a/subsys/bluetooth/host/gatt.c
+++ b/subsys/bluetooth/host/gatt.c
@@ -2012,8 +2012,9 @@ static u8_t update_ccc(const struct bt_gatt_attr *attr, void *user_data)
        ccc = attr->user_data;
 
        for (i = 0; i < ARRAY_SIZE(ccc->cfg); i++) {
-               /* Ignore configuration for different peer */
-               if (bt_conn_addr_le_cmp(conn, &ccc->cfg[i].peer)) {
+               /* Ignore configuration for different peer or not active */
+               if (!ccc->cfg[i].value ||
+                   bt_conn_addr_le_cmp(conn, &ccc->cfg[i].peer)) {
                        continue;
                }
 
@@ -2043,14 +2044,14 @@ static u8_t update_ccc(const struct bt_gatt_attr *attr, void *user_data)
                        }
                }
 
-               if (ccc->cfg[i].value) {
-                       gatt_ccc_changed(attr, ccc);
-                       if (IS_ENABLED(CONFIG_BT_GATT_SERVICE_CHANGED) &&
-                           ccc == &sc_ccc) {
-                               sc_restore(conn);
-                       }
-                       return BT_GATT_ITER_CONTINUE;
+               gatt_ccc_changed(attr, ccc);
+
+               if (IS_ENABLED(CONFIG_BT_GATT_SERVICE_CHANGED) &&
+                   ccc == &sc_ccc) {
+                       sc_restore(conn);
                }
+
+               return BT_GATT_ITER_CONTINUE;
        }
 
        return BT_GATT_ITER_CONTINUE;
```

Reply:
@Vudentz Your change seems less intrusive and achieves the same goal. I'll happily close this PR if you open one with your change.

Reply:
@wopu-ot Just update this one to that change?

Reply:
Yep, just makes this changes into your PR so we don't have to start a new one.

Reply:
Sorry for the delay, I updated the PR now.